index.rst

```diff
@@ -1,61 +1,333 @@
+
+Quotas will be implemented in the service API frontend.
+Usage information will be stored in the same SQL database used to store job metadata and used to make quota decisions.
+
+Throttling will be implemented the same way, using the same data.
+Rather than rejecting the request as with a quota limit, throttled requests may be set to a lower priority when dispatched via Dramatiq so that they will be satisfied only after higher-priority requests are complete.
+If we develop a mechanism for estimating the cost of a request, throttling may also reject expensive requests while allowing simple requests.
+
+If the service starts throttling, sync requests may not be satisfiable within a reasonable HTTP timeout interval.
+Therefore, depending on the severity of the throttling, the image cutout service may begin rejecting sync requests from a given user and requiring all requests be async.
+
+All of these decisions will be made by the API service layer when the user attempts to start a new job or makes a sync request.
+
+.. _cutout:
+
+Performing the cutout
+=====================
+
+To ensure the cutout operation is performed by properly-vetted scientific code, the image cutout will be done via a task.
+For some types of cutouts, such as cutouts from PVIs that must be reconstructed from raw images, this may be a pipeline task.
+
+Currently, pipeline tasks must be invoked via the command line, but the expectation is that pipelines will add a way of invoking a pipeline task via a Python API.
+Once that is available, each cutout worker can be a long-running Python process that works through a queue of cutout requests, without paying the cost of loading Python libraries and preparing supporting resources for each cutout action.
+
+.. _results:
+
+Results
+=======
+
+Result format
+-------------
+
+All cutout requests will create a FITS file stored in a new output Butler collection.
+The metadata about the request that would be returned as metadata for a UWS async job (see :ref:`task-storage`) will also be stored in that Butler collection so that the collection has the provenance of the cutouts.
+
+The primary output of a cutout operation in the initial implementation will be a single FITS file.
+Each filtering parameter produces a separate cutout.
+The cutout images will be stored as extensions in the result FITS file, not in the Basic FITS HDU.
+
+The result of a sync request that does not request an alternate image format is the FITS file.
+Therefore the sync API will redirect to the FITS file result of the underlying async job.
+
+The job representation for a successful async request will list at least two results: the FITS file, and the URL for the output Butler collection that contains both that FITS file and the metadata about the cutout request.
```

Comment:
Butler doesn't have a concept of a URL to a collection. It only has URLs to datasets.

In OCPS I think they work around this problem by calling `butler retrieve-artifacts` (or equivalent python API) to retrieve all the relevant output files and place them in a temporary location that is available as a URL.

Reply:
The implication of this is that the image cutout service cannot return a Butler collection, since the only thing that a UWS job can return is a set of URLs. This means that the user will not be able to access or move the Butler collection as a collection, only retrieve the individual files contained in it. I'll add that as an open question if that's okay, or if we'll need to add the concept of a URL to a Butler collection.

Reply:
We have a convention that the URL includes the run collection name in the root as a "directory" but that's just something that comes out of the file template we define and not something butler or datastore is requiring will happen. In theory the equivalent of `os.path.commonprefix` would return a bucket location that is what you need but as I say, that's pure chance at the moment (but might be true for ever).

Reply:
Note that this is primarily interesting in a client/server Butler future, since that's presumably what we will go into production with. So if there will be a concept of a URL to a collection with client/server Butler, that's all that's needed, and we can work around it until then. (So perhaps this is a design consideration for client/server Butler.)

Reply:
It's fundamentally the case that butler datastore doesn't know about collections at all. The registry knows about collections but datastore is just writing files using a template constructed from the DatasetRef that it's been told to write.  Client/server vs direct connection doesn't affect that. As I say above, we could enforce it solely by policy in requiring that every file template starts with a "run" and then from the set of matching URIs we could construct a common prefix that would effectively be the run collection URL. That's something a client could construct if butler itself was enforcing a template-starts-with-run policy.

Reply:
I think the key point that I'm missing is that I don't understand the API to the Butler and probably need to go do some reading. The mental concept I had was that a collection would be something a client could enumerate the members of and retrieve individual objects from, but it sounds like that mental model may be incorrect.

Reply:
A client can ask registry for all the datasets in a collection. The client can also ask datastore for the URIs for all the artifacts in those datasets. Armed with that knowledge a client could (if policy allowed) work out a common prefix for all those URIs that would in theory be a URI to the collection.  Butler though does not think of a URI to a collection as a concept itself.

Reply:
If the request to the registry for all the datasets in a collection is representable as a URL (which presumably it might be with client/server butler?), then maybe we could treat that as the URL to the collection? It at least seems like something reasonable to return as one of the results of an image cutout.

Reply:
BTW, I should mention that one of the points of making the Butler collection available as a collection rather than as a set of output files was so that someone could more easily transfer the entire Butler collection into their own space and out of the scratch space used for image cutout responses. So I think one of the goals is to provide whatever reference would provide the necessary inputs to `Butler.transfer_from`.

Reply:
Querying registry for all datasets in the collection is going to (1) need auth, and (2) retrieve a collection of DatsetRef in JSON form (with the URIs inside). We could add a new API for "return URIs (as signed URLs) for this query" but you still need to sort out the auth issue -- is the collection owned by the SODA service? Assuming the SODA service is only runnable by data rights users presumably you could set group ownership for the collection to that user or run the cutout service as that user directly? Is there any way for a user to say that they'd actually want the results to turn up directly in their personal user collection behind the scenes? Would you want the SODA output butler repo to be independent of the main repo? ie do we set up a chained datastore with main repo as input and a completely separate datastore as output and transfer relevant registry entries over -- this would let the services butler be completely independent and might have some administration pluses.

`Butler.transfer_from` takes `DatasetRef`, not URLs, because we need access to the registry information from the source butler.

Bit of a stream of consciousness this morning...

Reply:
> Assuming the SODA service is only runnable by data rights users presumably you could set group ownership for the collection to that user or run the cutout service as that user directly?

The SODA service will make client/server Butler calls as the user requesting the cutout, using a delegated access token. I'll make that explicit in this document.

> Is there any way for a user to say that they'd actually want the results to turn up directly in their personal user collection behind the scenes?

The SODA specification says this is possible but doesn't define any way to do it. It would therefore have to be a custom parameter, which we would either have to require (which may cause problems with generic SODA clients) or would still have to decide what happens if it's not provided.

> Would you want the SODA output butler repo to be independent of the main repo?

I have no opinion on this. I'm not sure what the trade-offs would be.

> (2) retrieve a collection of DatsetRef in JSON form (with the URIs inside)

I think this is okay. We can define the result of the Butler collection to be this JSON document and provide a URL that would retrieve it. If this is the input to `Butler.transfer_from`, that feels like the right approach. I don't see any obstacle to returning URLs that require authentication.